internal/ingress/controller/parser/parser.go

```diff
@@ -1115,14 +1116,87 @@ func cloneStringPointerSlice(array ...*string) []*string {
 	return res
 }
 
-func overrideUpstream(upstream *Upstream,
+func overrideUpstreamHostHeader(upstream *kong.Upstream, anns map[string]string) {
+	if upstream == nil {
+		return
+	}
+	host := annotations.ExtractHostHeader(anns)
+	if host == "" {
+		return
+	}
+	upstream.HostHeader = kong.String(host)
+}
+
+// overrideUpstreamByAnnotation modifies the Kong upstream based on annotations
+// on the Kubernetes service.
+func overrideUpstreamByAnnotation(upstream *kong.Upstream,
+	anns map[string]string) {
+	if upstream == nil {
+		return
+	}
+	overrideUpstreamHostHeader(upstream, anns)
+}
+
+// overrideUpstreamByKongIngress modifies the Kong upstream based on KongIngresses
+// associated with the Kubernetes service.
+func overrideUpstreamByKongIngress(upstream *Upstream,
 	kongIngress *configurationv1.KongIngress) {
-	if kongIngress == nil || kongIngress.Upstream == nil || upstream == nil {
+	if upstream == nil {
+		return
+	}
+
+	if kongIngress == nil || kongIngress.Upstream == nil {
+		return
+	}
+	u := kongIngress.Upstream
+	if u.HostHeader != nil {
+		upstream.HostHeader = kong.String(*u.HostHeader)
+	}
+	if u.Algorithm != nil {
+		upstream.Algorithm = kong.String(*u.Algorithm)
+	}
+	if u.Slots != nil {
+		upstream.Slots = kong.Int(*u.Slots)
+	}
+	if u.Healthchecks != nil {
+		upstream.Healthchecks = u.Healthchecks
+	}
+	if u.HashOn != nil {
+		upstream.HashOn = kong.String(*u.HashOn)
+	}
+	if u.HashFallback != nil {
+		upstream.HashFallback = kong.String(*u.HashFallback)
+	}
+	if u.HashOnHeader != nil {
+		upstream.HashOnHeader = kong.String(*u.HashOnHeader)
+	}
+	if u.HashFallback != nil {
+		upstream.HashFallback = kong.String(*u.HashFallback)
+	}
+	if u.HashFallbackHeader != nil {
+		upstream.HashFallbackHeader = kong.String(*u.HashFallbackHeader)
+	}
+	if u.HashOnCookie != nil {
+		upstream.HashOnCookie = kong.String(*u.HashOnCookie)
+	}
+	if u.HashOnCookiePath != nil {
+		upstream.HashOnCookiePath = kong.String(*u.HashOnCookiePath)
+	}
```

Comment:
Any reason for expanding these? Why can't we simply copy over the whole object?

Reply:
I'd originally tried to [just move the copy inside overrideUpstreamByKongIngress](https://gist.github.com/rainest/bd8b41e5643936528c46be1ec52dc47c#file-copy-diff), but this resulted in a [segfault when setting the name after](https://gist.github.com/rainest/bd8b41e5643936528c46be1ec52dc47c#file-segfault-txt-L14), so I cribbed from the closest equivalent (the service override, which splits it out).

Reading the code that interacts with the name a second time, I now understand more why it exists and why that occurred (name has to be assigned before the copy clobbers it). Moved it inside the KongIngress override since the annotation override can't touch it, and it makes more sense to read in sequence.